### DISCUSSION TITLE
Add reference coordinate system on patient

### DISCUSSION THREAD

**Discussion by martynakielbasa** [PlusToolkit/PlusLib#1007] [Q&A]
Hello again, 

I have encountered a problem that I am not able to solve on my own. 

I have performed a full calibration using fCal.exe and now I would like to record a tracked ultrasound sequence of a femur phantom. I want the tool on the phantom to be the reference coordinate system, so that I can move it, rotate it etc. while scanning (picture below). I have the .rom file of the tool.  
![313998369_529109098646890_4545712011462187833_n](https://user-images.githubusercontent.com/111685804/199518918-11bddc05-2159-4fab-a7f5-866909ad9264.jpg)
So far I haven't had much success in solving this issue. Here's what I did: in my config file (the one with the calibration result) I swapped the .rom files of the tools (the one on the calibration phantom and the one on the femur phantom). When I connected my devices, the tool on the femur phantom was seen as 'Reference'. I know that my reasoning was probably wrong, but I wanted to try out the easiest solution that came to my min

After volume reconstruction I got a shattered model, that's how I know the coordinate systems are not correctly specified. 

Could you please instruct me on how to solve the problem?
FYI, I'm attaching the config file (with swapped file). 
[config_file.txt](https://github.com/PlusToolkit/PlusLib/files/9921385/config_file.txt)

**Comment by Sunderlandkyl** [comment]
How are you reconstructing the volume, using Plus or 3D Slicer?

For Plus, I think you need to specify ReferenceCoordinateFrame and ImageCoordinateFrame in the VolumeReconstruction element in the config file. http://perk-software.cs.queensu.ca/plus/doc/nightly/user/AlgorithmVolumeReconstruction.html

**Comment by martynakielbasa** [comment]
I am using Plus' VolumeReconstructor.exe. I only use Slicer to view the volume. 

I was using the command-line parameter '--image-to-reference-transform=ImageToReference'. 
Now I have specified ReferenceCoordinateSystem as "Reference" and ImageCoordinateSystem as "Image" and the result of reconstruction is still incorrect. The picture below is what I got after reconstruction. It was supposed to be the femural head. 
![image](https://user-images.githubusercontent.com/111685804/199527786-e2bb510b-5f34-4b7c-bd14-70c7c3dba936.png)

**Comment by Sunderlandkyl** [comment]
When you visualize the tracked ultrasound slice in 3D Slicer, does the tracking seem good?

**Comment by martynakielbasa** [comment]
If you mean the three slices from different planes, then not really - I'm not able to see the correct shape of the bone.
![image](https://user-images.githubusercontent.com/111685804/199543871-17a4beec-6362-4a9c-8b1f-80c2bd1ac68f.png)

**Comment by Sunderlandkyl** [comment]
No, I meant the live tracked ultrasound image.

I haven't tested it, but this config file should stream the live ultrasound image to Slicer in the Reference coordinate system.
[stream_us_config_file.txt](https://github.com/PlusToolkit/PlusLib/files/9922639/stream_us_config_file.txt)

You'll need to use the "Volume reslice driver" module from SlicerIGT to display the slice as it's tracked.

**Comment by martynakielbasa** [comment]
I'll test it as soon as I get to the lab again and let you know about the the results.

**Comment by martynakielbasa** [comment]
The tracking of the slice seems good, I tried moving the probe in various directions and also rotate it a bit (result below). It gets a bit laggy but I assume that's when the camera can't see the markers on the probe. 

I'm attaching two clips: one from Slicer with the tracked ultrasound slice and another of irl prove movement:

https://drive.google.com/drive/folders/1yblxomDxxZpqb28U08CPKEtnXT3X_MOk?usp=share_link

**Comment by martynakielbasa** [comment]
I also tried to use Slicer to reconstruct volume, but I was not able to record anything (e.g. scout scan). I assumed that it was impossible due to the fact that there was no CaptureDevice available (I compared it with your exemplary elbow sweep file and noticed it then). Does it have something to do with the config file?

My file:
![Zrzut ekranu (1889)](https://user-images.githubusercontent.com/111685804/199961829-ca33676f-8356-4084-9482-a05a23bf54bd.png)

The elbow sweep file:
![Zrzut ekranu (1890)](https://user-images.githubusercontent.com/111685804/199961842-879865d0-c806-4830-8faa-3569a16e9716.png)

**Comment by Sunderlandkyl** [comment]
Yes, the scout scan needs a Virtual capture device. Probably you can just copy it from the other config file.

You can also perform live volume reconstruction directly in Slicer using the Volume Reconstruction module in SlicerIGT: https://www.youtube.com/watch?v=2vXeJxYFou4

**Comment by martynakielbasa** [comment]
I don't see the Volume Reconstruction module in SlicerIGT. 
I have Slicer 5.0.3. The SlicerIGT extension version is c73b307 (2022-07-08). 
![image](https://user-images.githubusercontent.com/111685804/200115149-c5b40b87-5988-4b1e-95ae-3a1bbb190479.png)

**Comment by Sunderlandkyl** [comment]
For Slicer 5.0.3, you'll need to install the SlicerIGSIO extension.

**Comment by martynakielbasa** [comment]
I tried using the Volume Reconstruction module from Slicer, but still got weird results. The reconstructed volume seems to be lacking depth and cramped. 

![image](https://user-images.githubusercontent.com/111685804/200326487-da0944ef-a775-461a-b4dd-2d0de17d0a7c.png)
![image](https://user-images.githubusercontent.com/111685804/200326497-502afb8b-81e2-461a-b51f-fe542e41177c.png)

I also noticed that the image from the ultrasound probe was not displayed correctly - it looked like it was squeezed a bit. I use the Telemed linear probe size 60 mm - maybe that's the problem?

I also tried changing the transformation hierarchy. I tried two different options - none of the worked correctly:
![image](https://user-images.githubusercontent.com/111685804/200327850-39a43643-01fe-4def-9933-553e32a5d015.png)
![image](https://user-images.githubusercontent.com/111685804/200327920-675531f6-cece-49f7-aee0-b0e71b2f5ad5.png)
Perhaps I still haven't used the proper one.

**Comment by Sunderlandkyl** [comment]
Image_Image transforming it with ImageToReference, and sending the image as Image_Reference are both correct.

Are you using the same ultrasound probe depth that you calibrated your probe with?

**Comment by martynakielbasa** [comment]
Hello again, 

found a solution to my problem by recalibrating with UF image orientation and adding another coordinate system named 'Patient' next to the existent 'Reference' one. I then changed the ReferenceCoordinateSystem for Volume Reconstruction and conducted live volume reconstruction using Slicer. Results below (not a full model yet, only a control scan):

![image](https://user-images.githubusercontent.com/111685804/201651883-e4ed03e6-3168-42e9-b1b5-52d31b730bbc.png)

Thank you for all your help and and involvement in this matter!

**Comment by lassoan** [comment]
Great, thanks for the update!
